src/Pirouette/SMT/Constraints.hs

```diff
@@ -25,50 +25,45 @@ import Pirouette.Term.Syntax
 
 data ConstraintSet lang meta = ConstraintSet
-  { -- | Represents the set of productive symbolic variable assignments we know of.
-    -- Productive in the sense that we will always have at least one constructor
-    -- on the rhs.
-    csAssignments :: M.Map meta (TermMeta lang meta),
-    -- | Represents the symbolic variable equivalences we know of. Satisfies the invariant
-    -- that if @M.lookup v csmetaEq == Just v'@, then @v' < v@ (lexicographically)
-    csMetaEq :: M.Map meta meta,
+  { -- | Represents the set of productive symbolic variable assignments we know
+    -- of. Productive in the sense that we will always have at least one
+    -- constructor on the rhs. The assignments are stored from /equivalence
+    -- classes of symbolic variables/ to @TermMeta@.
+    csAssignments :: UF.UnionFind meta (TermMeta lang meta),
     -- | Any potential native constraint we might need.
     csNative :: [PureSMT.SExpr],
     -- | And finally, pairs of terms we've discovered /not/ to be equal
     csRelations :: [(TermRelation, TermMeta lang meta, TermMeta lang meta)]
   }
 
-instance (Ord meta) => Semigroup (ConstraintSet lang meta) where
-  s1 <> s2 =
-    ConstraintSet
-      (csAssignments s1 <> csAssignments s2)
-      (csMetaEq s1 <> csMetaEq s2)
-      (csNative s1 <> csNative s2)
-      (csRelations s1 <> csRelations s2)
+-- instance (Ord meta, Semigroup meta) => Semigroup (ConstraintSet lang meta) where
+--   s1 <> s2 =
+--     ConstraintSet
+--       (csAssignments s1 <> csAssignments s2)
+--       (csNative s1 <> csNative s2)
+--       (csRelations s1 <> csRelations s2)
 
-instance (Ord meta) => Monoid (ConstraintSet lang meta) where
-  mempty = def
+-- instance (Ord meta) => Monoid (ConstraintSet lang meta) where
+--   mempty = def
 
 instance Default (ConstraintSet lang meta) where
-  def = ConstraintSet M.empty M.empty [] []
+  def = ConstraintSet UF.empty [] []
 
 instance (LanguagePretty lang, Pretty meta) => Show (ConstraintSet lang meta) where
   show = show . pretty
 
-instance (LanguagePretty lang, Pretty meta) => Pretty (ConstraintSet lang meta) where
-  pretty ConstraintSet {..} =
-    vsep $
-      map (uncurry prettyAssignment) (M.toList csAssignments)
-        ++ map prettyRelations csRelations
-        ++ map (uncurry prettyMetaEq) (M.toList csMetaEq)
-        ++ map prettyNative csNative
-    where
-      prettyAssignment v t = pretty v <+> ":=" <+> pretty t
-      prettyMetaEq v u = pretty v <+> "~~" <+> pretty u
-      prettyRelations (r, v, u) = pretty v <+> pretty r <+> pretty u
-      prettyNative n = pretty (show n)
+instance (Pretty meta, Pretty (TermMeta lang meta)) => Pretty (ConstraintSet lang meta) where
+  pretty ConstraintSet {..} = undefined
+    -- vsep $
+    --   [pretty csAssignments]
+    --     ++ map prettyRelations csRelations
+    --     ++ map prettyNative csNative
+    -- where
+    --   prettyRelations (r, v, u) = pretty v <+> pretty r <+> pretty u
+    --   prettyNative n = pretty (show n)
```

Comment:
I was able to get it to work by basically copying the pretty code in the `UnionFind` module. Otherwise, I was only able to get it to work by adding `Pretty (UF.UnionFind meta (TermMeta lang meta))` or similar in lots of places, not just in the `Constraints` file

Reply:
Oh, nice! If you have a version that works, feel free to push it here! I would have expected `Pretty meta` and `Pretty (TermMeta lang meta)` to be exactly the constraints necessary to call `pretty` on a `UnionFind meta (TermMeta lang meta)`, but apparently not, and I don't understand why not.

Reply:
I pushed up a version that works, though as I mentioned it kind of copies the pretty code in the `UnionFind` module. If it was unclear, I only needed to add these extra constraints if I was trying to use the pretty instance from `UnionFind` instead of copying it, and I don't know why that is. I'll do some more tinkering tomorrow now that some code changed to see if I can figure it out